### PR TITLE
KAFKA-10588 Rename kafka-console-consumer CLI command line arguments for KIP-629

### PR DESCRIPTION
**Pull request by OmniaGM**
This PR marks `--whitelist` as deprecated argument and introduce `--include` for `kafka-console-consumer`

Note:
There is no test using `--whitelist`
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by OmniaGM** [comment]
Hi, @xvrl, @gwenshap can you please have a look into this? Thanks

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: small typo
```suggestion
      "Regular expression specifying list of topics to include for consumption.")
```

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: nit, since it's not really a list, maybe we should use the name `include` or `includeTopics` instead, similar to what you did here https://github.com/apache/kafka/commit/51796bcdef2e93ee4176415961ab98dd7407ebec#diff-54be3f408ff6ce6c9fb454bd25a3c67b1dea23b5da663f9d7e17b5b946222324R101 ?

**Review by xvrl — APPROVED** [review]
LGTM, if @gwenshap or another Kafka committer can approve that'd be great

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I would still mention `whitelist` here as it is still usable.

**Review by dajac — COMMENTED** [review]
@OmniaGM Thanks for the PR. I left a small comment. Could we also add test(s) in `ConsoleConsumerTest` to cover the new argument?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I'm not sure if it makes sense to list deprecated configurations in error messages, do we do this elsewhere? The option would still show up in the usage help that gets printed following the error message.

Comment (dajac) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I don't know if we did this somewhere else.

I just think that the error message is confusing for folks who would use the deprecated argument. Say you provide the deprecated one with topic and you get an error which mention the new one. This is confusing.

As an alternative, we could actually mention the one that was used in the error message.

Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala:326]: We can have `include/whitelist/topic` and remove it when we completely remove `--whitelist` later.

Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I agree with @xvrl it doesn't make sense to list deprecated config in an error message. I can add an extra note that `whitelist` is deprecated. This error show if a user is setting `whitelist` or `include` with `topic` which already a wrong behaviour so listing the deprecated config in the error message doesn't make sense either.

Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: The error message now will be `Exactly one of include/topic is required. whitelist is DEPRECATED use --include instead.` if there's any filter specified (`whitelist` or `include`) with `topic` or if the user didn't set value for both filter options and topic.

Also If a user set value for both `whitelist` and `include` there will be a warning message `--whitelist is DEPRECATED and ignored if --include specified.`

**Comment by OmniaGM** [comment]
> @OmniaGM Thanks for the PR. I left a small comment. Could we also add test(s) in `ConsoleConsumerTest` to cover the new argument?

I added few test cases to cover both new and old ones.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: maybe we can check `options.has(whitelistOpt)` here and only print the deprecation message if they provide the deprecated argument?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: nit: The parenthesis are not necessary. I would remove them in order to be consistent with the other arguments.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: Checking this is a very good idea. However, I would `printUsageAndDie` when both are used together instead of ignoring `--whitelist`. We could say `Options --include and --whitelist cannot be specified together. --whitelist is DEPRECATED, use --include instead.`.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: nit: Should we prefix all the options with `--` to be consistent in the error message?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala:195]: We could remove `ignored if --include specified` if we don't accept both together.

File: core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala
Comment (dajac) [inline core/src/test/scala/unit/kafka/tools/ConsoleConsumerTest.scala:168]: This test must be updated if we don't accept both options together.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: nit: Would it be simpler to check `topicArg != null && includedTopicsArg != null` in the `if`?

**Review by dajac — COMMENTED** [review]
@OmniaGM Thanks for the update. I left a few small comments. We are good to go once they are addressed.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala:195]: I think with some other deprecated arguments we simply ignore the deprecated if the new one specified. for example 
- `--broker-list` and `--bootstrap-server` in ConsoleProducer, VerifiableConsumer, VerifiableProducer and GetOffsetShell
And the scripts doesn't fail if both deprecated and new one existed.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: @dajac not sure if we should do this. With other deprecated arguments, we simply ignore the deprecated one, if the new one specified. for example I can see that `--broker-list` is simple ignored if `--bootstrap-server` specified in `ConsoleProducer`, `VerifiableConsumer`, `VerifiableProducer` and `GetOffsetShell`. And the scripts don't fail if both deprecated and new one existed. So to be consistent with this deprecation approach maybe we should follow the same. Do we fail on other deprecated arguments if they specified with the new arguments?

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: The error is fired if both topic and filter specified (which makes topicOrFilterOptArgs.size = 2) or if neither topic or filter is specified which makes (topicOrFilterOptArgs.size = 0) Not only if both aren't null. 
Maybe to make it more readable we can 

```
noTopicOrFilterSpecified = topicArg == null && includedTopicsArg == null
bothTopicAndFilterSpecified = topicArg != null && includedTopicsArg != null

if (noTopicOrFilterSpecified || bothTopicAndFilterSpecified) {
//handle error.
}
```

Comment (dajac) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: Ah right. I missed the second case. My suggestion does not make sense then. Please, ignore it.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (dajac) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I know. I don't think that we have been consistent about this but this does not prevent us from doing better here :). I am not a fan of having this warning printed out all the time for existing users. It might even break existing usages of the tool if someone does not expect it this new line.

Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I think if someone started to use `--include` they already noticed the usage help message which explained the ignoring part. But I got what you are saying. The other point is we already merged MirrorMaker script with only ignoring in this pr #10937 do we need to raise another one to handle this in the same way for consistency? (keeping in mind that  KIP-629 suppose to be part of 3.0.0 release and the code freeze is tomorrow July 14th, 2021)

Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: The other option is to have another Jira ticket to change the ignoring for all deprecated arguments in case of the existence of the new ones to be handled probably later for another release.

Comment (dajac) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I see. In this case, I would remove that warning and ignore `whitelist` if `include` is provided as you did in the first place to stay consistent with the other tools.

My only ask was to extend the error message to mention `--whitelist` if used like you did here:
```
CommandLineUtils.printUsageAndDie(parser, s"Exactly one of --include/--topic is required. " +
        s"${if (options.has(whitelistOpt)) "--whitelist is DEPRECATED use --include instead."}")
```

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [inline core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: you could also write this as an XOR `(topicArg != null ^ includedTopicsArg != null)`, and add a comment, not sure which one is more readable.

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: agree with staying consistent with other tools ignoring deprecated args if that's what we do elsewhere

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (xvrl) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala:195]: we can probably resolve this, since we decided to ignore deprecated args

File: core/src/main/scala/kafka/tools/ConsoleConsumer.scala
Comment (OmniaGM) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I revert it back to 
```
val topicOrFilterArgs = List(topicArg, includedTopicsArg).filterNot(_ == null)
if (topicOrFilterArgs.size != 1) {
//handle error
}
```
And added a comment as well is seems more readable than the XOR and my previous handling

Comment (xvrl) [reply core/src/main/scala/kafka/tools/ConsoleConsumer.scala]: I don't have a strong preference, feel free to use whichever approach seems more readable to you

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Comment by dajac** [comment]
@xvrl Have we already discussed with @kkonstantine to still include this in 3.0?

**Comment by xvrl** [comment]
@dajac yes, we want to include this in 3.0

**Comment by dajac** [comment]
@xvrl Ack. I will merge as soon as we get a clean build for it. CI seems to be stuck at the moment...

**Comment by dajac** [comment]
Merged to trunk and to 3.0. cc @kkonstantine

**Pull request merged**
